src/codegen.cpp

```diff
@@ -3781,14 +3837,19 @@ static void emit_assignment(jl_value_t *l, jl_value_t *r, jl_codectx_t *ctx)
                     emit_unionmove(dest, slot, isboxed, false, NULL, ctx);
                 Value *gcroot = NULL;
                 if (isboxed) {
-                    if (slot.gcroot)
+                    Value *box;
+                    if (slot.gcroot) {
                         gcroot = emit_local_root(ctx);
-                    else
+                        box = builder.CreateLoad(slot.gcroot);
```

Comment:
`load(gcroot) != boxed(slot)` in general. in particular, slot could represent a gep (getfield) or bitcast.